tools/src/main/java/org/apache/kafka/tools/VerifiableConsumer.java

```diff
@@ -0,0 +1,609 @@
+
+        List<RecordSetSummary> summaries = new ArrayList<>();
+        for (TopicPartition tp : records.partitions()) {
+            List<ConsumerRecord<String, String>> partitionRecords = records.records(tp);
+
+            if (hasMessageLimit() && consumedMessages + partitionRecords.size() > maxMessages)
+                partitionRecords = partitionRecords.subList(0, maxMessages - consumedMessages);
+
+            if (partitionRecords.isEmpty())
+                continue;
+
+            long minOffset = partitionRecords.get(0).offset();
+            long maxOffset = partitionRecords.get(partitionRecords.size() - 1).offset();
+
+            offsets.put(tp, new OffsetAndMetadata(maxOffset+1));
+            summaries.add(new RecordSetSummary(tp.topic(), tp.partition(),
+                    partitionRecords.size(), minOffset, maxOffset));
+
+            if (verbose) {
+                for (ConsumerRecord<String, String> record : partitionRecords)
+                    printJson(new RecordData(record));
+            }
+
+            consumedMessages += partitionRecords.size();
+            if (isFinished())
+                break;
+        }
+
+        printJson(new RecordsConsumed(records.count(), summaries));
+        return offsets;
+    }
+
+    @Override
+    public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+        List<CommitData> committedOffsets = new ArrayList<>();
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> offsetEntry : offsets.entrySet()) {
+            TopicPartition tp = offsetEntry.getKey();
+            committedOffsets.add(new CommitData(tp.topic(), tp.partition(), offsetEntry.getValue().offset()));
+        }
+
+        boolean success = true;
+        String error = null;
+        if (exception != null) {
+            success = false;
+            error = exception.getMessage();
+        }
+        printJson(new OffsetsCommitted(committedOffsets, error, success));
+    }
+
+    @Override
+    public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        printJson(new PartitionsAssigned(partitions));
+    }
+
+    @Override
+    public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        printJson(new PartitionsRevoked(partitions));
+    }
+
+    private void printJson(Object data) {
+        try {
+            out.println(mapper.writeValueAsString(data));
+        } catch (JsonProcessingException e) {
+            out.println("Bad data can't be written as json: " + e.getMessage());
+        }
+    }
+
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        try {
+            consumer.commitSync(offsets);
+            onComplete(offsets, null);
+        } catch (WakeupException e) {
+            // we only call wakeup() once to close the consumer, so this recursion should be safe
+            commitSync(offsets);
+            throw e;
```

Comment:
The wakeup exception is needed to break from the poll loop. We could alternatively set a closed flag.